### PR TITLE
provisioner: multinode motr provisioner support added

### PR DESCRIPTION
**Pull request by bansisoni**
- EOS-16363
- lnet sanity test added for multinode cluster
- lvm creation improved, so not fail if volume group exist
- general error handling of motr_setup commands
- motr_cfg.sh fixed to run without m0provision and salt
- always return 0 on success
- failure will retun with error code and traceback

Signed-off-by: Bansidhar Soni <bansidhar.soni@seagate.com>

### CODE REVIEW

**Comment by bansisoni** [comment]
Help msg:
```[root@ssc-vm-1399 cortx-motr]# URL0=json:///home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/conf/examples/motr_configV1.json            

[root@ssc-vm-1399 cortx-motr]# motr_setup --help
usage: Motr Setup command [-h]
                          {backup,cleanup,config,init,post_install,reset,restore,test,upgrade}
                          ...

positional arguments:
  {backup,cleanup,config,init,post_install,reset,restore,test,upgrade}
    backup              setup BackupCmd
    cleanup             setup CleanupCmd
    config              setup ConfigCmd
    init                setup InitCmd
    post_install        setup PostInstallCmd
    reset               setup ResetCmd
    restore             setup RestoreCmd
    test                setup TestCmd
    upgrade             setup UpgradeCmd

optional arguments:
  -h, --help            show this help message and exit

[root@ssc-vm-1399 cortx-motr]# motr_setup init --help
usage: Motr Setup command init [-h] [--config CONFIG] [--debug]
                               [args [args ...]]

positional arguments:
  args             args

optional arguments:
  -h, --help       show this help message and exit
  --config CONFIG  Config URL
  --debug, -d      enable debug print
  ```

**Comment by bansisoni** [comment]
Success msg:
```
[root@ssc-vm-1399 cortx-motr]# motr_setup post_install --config $URL0
Node name: srvnode-1
Processing post_install json:///home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/conf/examples/motr_configV1.json []
Checking for /lib/modules/3.10.0-1127.el7.x86_64/kernel/fs/motr/m0tr.ko
Checking for /etc/sysconfig/motr
SUCCESS
[root@ssc-vm-1399 cortx-motr]# echo $?
0

[root@ssc-vm-1399 cortx-motr]# motr_setup config --config $URL0
Node name: srvnode-1
Processing config json:///home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/conf/examples/motr_configV1.json []
SUCCESS
[root@ssc-vm-1399 cortx-motr]# echo $?
0

[root@ssc-vm-1399 cortx-motr]# motr_setup init --config $URL0
Node name: srvnode-1
Processing init json:///home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/conf/examples/motr_configV1.json []

lvm metadata_devices: ['/dev/sdb', '/dev/sdc']

metadata device: /dev/sdb
Checking for /etc/fstab
Checking for /dev/sdb
Removing vg_srvnode-1_md1 volume group
Creating physical volume from /dev/sdb
Creating vg_srvnode-1_md1 volume group from /dev/sdb
Adding srvnode-1 tag to vg_srvnode-1_md1 volume group
Scanning volume group
Creating lv_main_swap1 lvm from vg_srvnode-1_md1
Creating lv_raw_md1 lvm from vg_srvnode-1_md1
Make swap of /dev/vg_srvnode-1_md1/lv_main_swap1
Test /dev/vg_srvnode-1_md1/lv_main_swap1 swap device
Adding /dev/vg_srvnode-1_md1/lv_main_swap1 swap device to /etc/fstab
Swap entry added: /dev/vg_srvnode-1_md1/lv_main_swap1    swap    swap    defaults        0 0

metadata device: /dev/sdc
Checking for /etc/fstab
Checking for /dev/sdc
Removing vg_srvnode-1_md2 volume group
Creating physical volume from /dev/sdc
Creating vg_srvnode-1_md2 volume group from /dev/sdc
Adding srvnode-1 tag to vg_srvnode-1_md2 volume group
Scanning volume group
Creating lv_main_swap2 lvm from vg_srvnode-1_md2
Creating lv_raw_md2 lvm from vg_srvnode-1_md2
Make swap of /dev/vg_srvnode-1_md2/lv_main_swap2
Test /dev/vg_srvnode-1_md2/lv_main_swap2 swap device
Adding /dev/vg_srvnode-1_md2/lv_main_swap2 swap device to /etc/fstab
Swap entry added: /dev/vg_srvnode-1_md2/lv_main_swap2    swap    swap    defaults        0 0

lnet config: options lnet networks=tcp(eth1) config_on_load=1  lnet_peer_discovery_disabled=1
Restarting lnet service
SUCCESS
[root@ssc-vm-1399 cortx-motr]# echo $?
0

[root@ssc-vm-1399 cortx-motr]# motr_setup test --config $URL0
Node name: srvnode-1
Processing test json:///home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/conf/examples/motr_configV1.json []
rpm found: kmod-lustre-client
rpm found: lustre-client
lnet interface found: eth1
lnet interface ip: 192.168.5.84
ping on: 192.168.5.84
lnet pinging on all nodes in cluster
motr_setup init MUST be performed on all nodes before executing this
lctl ping on: 192.168.5.84@tcp
SUCCESS
[root@ssc-vm-1399 cortx-motr]# echo $?
0

```

**Comment by bansisoni** [comment]
Failure msg:
```
#failure case: device not found
[root@ssc-vm-1399 cortx-motr]# motr_setup init --config $URL0
Node name: srvnode-1
Processing init json:///home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/conf/examples/motr_configV1.json []

lvm metadata_devices: ['/dev/sdbb', '/dev/sdc']

metadata device: /dev/sdbb
Checking for /etc/fstab
Checking for /dev/sdbb

error[2]: /dev/sdbb not exist

Traceback (most recent call last):
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 189, in main
    command.process()
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 119, in process
    config_lvm(self)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py", line 332, in config_lvm
    create_lvm(self, metadata_devices.index(device), device)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py", line 272, in create_lvm
    validate_file(metadata_dev)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py", line 99, in validate_file
    raise MotrError(errno.ENOENT, "{} not exist".format(file))
motr_mini_prov.MotrError: error[2]: /dev/sdbb not exist

usage: /home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup [-h] <cmd> --config <url> <args>
where:
cmd           post_install, config, init, test, reset, cleanup, upgarde, backup, restore
url           Config URL
[root@ssc-vm-1399 cortx-motr]# echo $?
22

#failure case: machine-id not found
[root@ssc-vm-1399 cortx-motr]# motr_setup post_install --config $URL0

error[22]: Current node not found

Traceback (most recent call last):
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py", line 79, in get_current_node
    current_node = Conf.get(self._index, 'cluster>server_nodes')[machine_id]
KeyError: '0a31807456be14965f31b49314f5664e'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 188, in main
    command = Cmd.get_command(desc, argv[1:])
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 72, in get_command
    return args.command(args)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 91, in __init__
    super().__init__(args)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 39, in __init__
    self._server_id = get_current_node(self)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py", line 81, in get_current_node
    raise MotrError(errno.EINVAL, "Current node not found")
motr_mini_prov.MotrError: error[22]: Current node not found

usage: /home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup [-h] <cmd> --config <url> <args>
where:
cmd           post_install, config, init, test, reset, cleanup, upgarde, backup, restore
url           Config URL

```

**Comment by bansisoni** [comment]
Debug msg:
```
#debug print(it will print executed command output)
[root@ssc-vm-1399 cortx-motr]# motr_setup post_install --config $URL0 -d
[CMD] cat /etc/machine-id
[OUT]
0a31807456be14965f31b49314f5664e

[RET] 0
Node name: srvnode-1
Processing post_install json:///home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/conf/examples/motr_configV1.json []
[CMD] uname -r
[OUT]
3.10.0-1127.el7.x86_64

[RET] 0
Checking for /lib/modules/3.10.0-1127.el7.x86_64/kernel/fs/motr/m0tr.ko
Checking for /etc/sysconfig/motr
SUCCESS
[root@ssc-vm-1399 cortx-motr]# echo $?
0
```

**Comment by bansisoni** [comment]
No URL:
```
[root@ssc-vm-1399 cortx-motr]# motr_setup test --config
usage: Motr Setup command test [-h] [--config CONFIG] [--debug]
                               [args [args ...]]
Motr Setup command test: error: argument --config: expected one argument
[root@ssc-vm-1399 cortx-motr]# motr_setup test

error(22): Invalid URL None

Traceback (most recent call last):
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 188, in main
    command = Cmd.get_command(desc, argv[1:])
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 72, in get_command
    return args.command(args)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 128, in __init__
    super().__init__(args)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 36, in __init__
    Conf.load(self._index, self._url)
  File "/usr/lib/python3.6/site-packages/cortx/utils/conf_store/conf_store.py", line 172, in load
    Conf._conf.load(index, url)
  File "/usr/lib/python3.6/site-packages/cortx/utils/conf_store/conf_store.py", line 64, in load
    kv_store = KvStoreFactory.get_instance(kvs_url, self._delim)
  File "/usr/lib/python3.6/site-packages/cortx/utils/kv_store/kv_store.py", line 120, in get_instance
    raise KvError(errno.EINVAL, "Invalid URL %s", store_url)
cortx.utils.kv_store.error.KvError: error(22): Invalid URL None

usage: /home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup [-h] <cmd> --config <url> <args>
where:
cmd           post_install, config, init, test, reset, cleanup, upgarde, backup, restore
url           Config URL
```

**Comment by bansisoni** [comment]
lctl ping fail:
```
[root@ssc-vm-1399 cortx-motr]# motr_setup test --config $URL
Node name: srvnode-1
Processing test json:///home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/conf/examples/motr_prov_conf.json []
rpm found: kmod-lustre-client
rpm found: lustre-client
lnet interface found: eth1
lnet interface ip: 192.168.5.84
ping on: 192.168.5.84
lnet pinging on all nodes in cluster
motr_setup init MUST be performed on all nodes before executing this
lctl ping on: 192.168.5.84@tcp
lctl ping on: 192.168.47.253@tcp

error[1]: "lctl ping 192.168.47.253@tcp" command execution failed

Traceback (most recent call last):
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 189, in main
    command.process()
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 132, in process
    test_lnet(self)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py", line 443, in test_lnet
    lnet_ping(self)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py", line 413, in lnet_ping
    execute_command(self, cmd)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py", line 58, in execute_command
    raise MotrError(ps.returncode, f"\"{cmd}\" command execution failed")
motr_mini_prov.MotrError: error[1]: "lctl ping 192.168.47.253@tcp" command execution failed

usage: /home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup [-h] <cmd> --config <url> <args>
where:
cmd           post_install, config, init, test, reset, cleanup, upgarde, backup, restore
url           Config URL
[root@ssc-vm-1399 cortx-motr]# echo $?
22
```

**Comment by bansisoni** [comment]
Wrong lnet interface :
```
[root@ssc-vm-1399 cortx-motr]# motr_setup init --config $URL0
Node name: srvnode-1
Processing init json:///home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/conf/examples/motr_configV1.json []

lvm metadata_devices: ['/dev/sdb', '/dev/sdc']

metadata device: /dev/sdb
.
.
.
Adding /dev/vg_srvnode-1_md2/lv_main_swap2 swap device to /etc/fstab
Swap entry added: /dev/vg_srvnode-1_md2/lv_main_swap2    swap    swap    defaults        0 0

Validate private_interfaces[0]: eth3

error[1]: "ip addr show eth3" command execution failed

Traceback (most recent call last):
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 189, in main
    command.process()
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup", line 120, in process
    configure_net(self)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py", line 148, in configure_net
    configure_lnet(self)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py", line 168, in configure_lnet
    execute_command(self, cmd)
  File "/home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py", line 58, in execute_command
    raise MotrError(ps.returncode, f"\"{cmd}\" command execution failed")
motr_mini_prov.MotrError: error[1]: "ip addr show eth3" command execution failed

usage: /home/702612/shared/mini_prov5/cortx-motr/scripts/install/opt/seagate/cortx/motr/bin/motr_setup [-h] <cmd> --config <url> <args>
where:
cmd           post_install, config, init, test, reset, cleanup, upgarde, backup, restore
url           Config URL
[root@ssc-vm-1399 cortx-motr]# echo $?
22
```

File: scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py [hunk 1]
```diff
@@ -22,184 +22,321 @@
...
+        cmd = f"systemctl stop {service}"
+        execute_command(self, cmd)
+        cmd = f"systemctl start {service}"
+        execute_command(self, cmd)
+        cmd = f"systemctl status {service}"
+        execute_command(self, cmd)
 
 def validate_file(file):
     if not os.path.exists(file):
         raise MotrError(errno.ENOENT, "{} not exist".format(file))
```
Comment (huanghua78) [inline scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py:99]: Here, please use python f-string.

If F-String is used elsewhere in this python script, it's better to made them all use the same style.

File: scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py [hunk 2]
```diff
@@ -22,184 +22,321 @@
...
+    kernel_ver = op.replace('\n', '')
+    check_type(kernel_ver, str, "kernel version")
+
+    kernel_module = f"/lib/modules/{kernel_ver}/kernel/fs/motr/m0tr.ko"
+    sys.stdout.write(f"Checking for {kernel_module}\n")
+    validate_file(kernel_module)
+
+    sys.stdout.write(f"Checking for {MOTR_SYS_FILE}\n")
+    validate_file(MOTR_SYS_FILE)
+
```
Comment (huanghua78) [inline scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py:130]: Why is the "try except" block removed?

It is a best practice to have "try except".

Comment (bansisoni) [reply scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py:130]:
```
def main(argv: dict):
    try:	    try:
        desc = "Motr Setup command"	        desc = "Motr Setup command"
        command = Cmd.get_command(desc, argv[1:])	        command = Cmd.get_command(desc, argv[1:])
        command.process()	        command.process()


    except Exception as e:	    except Exception as e:
        sys.stderr.write("error: %s\n\n" % str(e))	        sys.stderr.write(f"\n{str(e)}\n\n")
        sys.stderr.write("%s\n" % traceback.format_exc())	        sys.stderr.write(f"{traceback.format_exc()}\n")
        Cmd.usage(argv[0])	        Cmd.usage(argv[0])
        return errno.EINVAL	        return errno.EINVAL
```
exception caught here in main.

File: scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py [hunk 3]
```diff
@@ -22,184 +22,321 @@
...
-        execute_command(cmd, TIMEOUT_SECS)
+    sys.stdout.write(f"Creating physical volume from {metadata_dev}\n")
+    cmd = f"pvcreate {metadata_dev} --yes"
+    execute_command(self, cmd)
 
-        cmd = f"vgchange --addtag {node_name} {vg_name}"
-        execute_command(cmd, TIMEOUT_SECS)
+    sys.stdout.write(f"Creating {vg_name} volume group from {metadata_dev}\n")
+    cmd = f"vgcreate {vg_name} {metadata_dev}"
+    execute_command(self, cmd)
```
Comment (huanghua78) [inline scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py:303]: Why are these commands taken out of "try ... except ..." block?

Comment (bansisoni) [reply scripts/install/opt/seagate/cortx/motr/bin/motr_mini_prov.py:303]: exception are raised intentionally to break execution if something fails. All, exception are handled in main function of motr_setup. which reports final error and call-trace of failure.
Only few exception are handled in motr_mini_prov.py which needs to handle.
Example:
```
  try:
        cmd = f"vgs {vg_name}"
        execute_command(self, cmd)
    except MotrError:
        pass
    else:
        sys.stdout.write(f"Removing {vg_name} volume group\n")

        del_swap_fstab_by_vg_name(self, vg_name)

        cmd = f"vgchange -an {vg_name}"
        execute_command(self, cmd)

        cmd = f"vgremove {vg_name} -ff"
        execute_command(self, cmd)
```

**Review by yeshpal-jain-cs — APPROVED** [review]

**Pull request merged**